docs/development/best-practices.md

```diff
@@ -0,0 +1,112 @@
+
+### Iterating objects & containers
+
+Use `for ( ... of ...)` loops instead of `[Array|Set|Map].prototype.forEach` and `for ( ... in ...)`.
+
+- Using `for ( ... in ...)` for objects is error-prone. It will include enumerable properties from the prototype chain
+- Using `for ( ... in ...)` to iterate over arrays, will counterintuitively give the array's indices
+- Avoid `[Array|Set|Map].prototype.forEach`. It makes code harder to debug and defeats some useful compiler checks like
+  reachability
+
+Only use `Array.prototype.map()` when the return value is used, otherwise use `for ( ... of ...)`.
+
+[Source](https://google.github.io/styleguide/tsguide.html#iterating-objects)
+
+### Exports
+
+Use named exports in all code.
+Avoid default `exports`.
+This way all `imports` follow the same pattern.
+
+[Source, reasoning and examples.](https://google.github.io/styleguide/tsguide.html#exports)
+
+### Imports
+
+Use ES6 module syntax.
+Avoid `require` `(as in import x = require('...');)` for imports.
+
+### HTTP & RESTful API request handling
+
+Prefer using [`got`](https://www.npmjs.com/package/got) for simplifying HTTP request handling.
+
+### Date & Time handling
+
+Use [`Luxon`](https://www.npmjs.com/package/luxon) for date and time handling.
+Use `UTC` to be time zone independent.
+
+[Example](https://github.com/renovatebot/renovate/blob/5043379847818ac1fa71ff69c098451975e95710/lib/modules/versioning/distro.ts#L133-L134)
+
+### Unit Testing
+
+- Use `it.each` rather than `test.each`
+- Mock Date/Time when testing a Date/Time dependent module
+  - For `Luxon` mocking see [Example](https://github.com/renovatebot/renovate/blob/5043379847818ac1fa71ff69c098451975e95710/lib/modules/versioning/distro.spec.ts#L7-L10)
+- Prefer `jest.spyOn` for mocking single functions, or mock entire modules
+  - Avoid overwriting functions, for example: (`func = jest.fn();`)
+
+### Fixtures
+
+- use `Fixture` class for loading fixtures
+
+```js
+Fixture.get('./file.json'); // for loading raw data
+Fixture.getJson('./file.json'); // for loading and parsing objects
+```
+
+### Working with vanilla JS files (renovate/tools only)
+
+Use [JSDoc](https://jsdoc.app/index.html) to declare types and function prototypes.
+
+[Example](https://github.com/renovatebot/renovate/blob/5043379847818ac1fa71ff69c098451975e95710/tools/distro-json-generate.mjs#L7-L17)
+
+### Classes
+
+Use [Typescript getter setters (Accessors) when needed](https://google.github.io/styleguide/tsguide.html#properties-used-outside-of-class-lexical-scope)
+. getter must be a `pure function`.
+
+- Omit constructors when defining Static classes
+- [No `#private` fields](https://google.github.io/styleguide/tsguide.html#private-fields). instead, use TypeScript's
+  visibility annotations.
+- Avoid underscore suffixes or prefixes, for example: `_prop`, use [whole words](https://google.github.io/styleguide/tsguide.html#properties-used-outside-of-class-lexical-scope) as
+  suffix/prefix i.e `internalProp`
```

Comment:
if we do plan to have this, do we want to permit some renovate related prefix for API scheme checking
like in here:
https://github.com/renovatebot/renovate/blob/b87b3b0087dbf4ca3dfb1794696b0e0259c1c765/lib/modules/versioning/index.spec.ts#L110-L113

as of now, underscore is used to indicate private class variables and methods (like _parse in genericVersioning?)

Reply:
We should probably be relaxed here, as we otherwise can't easily distinguish between private and public api 🤔

Reply:
so we leave the test as is.
should we keep this bullet point?